src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -0,0 +1,532 @@
+												ShardPlacementAccessType accessType);
+static void RecordParallelRelationAccess(Oid relationId, ShardPlacementAccessType
+										 placementAccess);
+static void RecordParallelRelationAccessToCache(Oid relationId,
+												ShardPlacementAccessType placementAccess);
+
+
+/*
+ * Empty RelationAccessHash, without destroying the hash table itself.
+ */
+void
+ResetRelationAccessHash()
+{
+	hash_delete_all(RelationAccessHash);
+}
+
+
+/*
+ * Allocate RelationAccessHash.
+ */
+void
+AllocateRelationAccessHash(void)
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(RelationAccessHashKey);
+	info.entrysize = sizeof(RelationAccessHashEntry);
+	info.hash = tag_hash;
+	info.hcxt = ConnectionContext;
+	hashFlags = (HASH_ELEM | HASH_BLOBS | HASH_CONTEXT);
+
+	RelationAccessHash = hash_create("citus connection cache (relationid)",
+									 8, &info, hashFlags);
+}
+
+
+/*
+ * AssociatePlacementAccessWithRelation associates the placement access to the
+ * distributed relation that the placement belongs to.
+ */
+void
+AssociatePlacementAccessWithRelation(ShardPlacement *placement,
+									 ShardPlacementAccessType accessType)
+{
+	uint64 shardId = INVALID_SHARD_ID;
+	Oid relationId = InvalidOid;
+
+	if (!ShouldRecordRelationAccess())
+	{
+		return;
+	}
+
+	shardId = placement->shardId;
+	relationId = RelationIdForShard(shardId);
+
+	/*
+	 * If a relation is partitioned, record accesses to all of its partitions as well.
+	 * We prefer to use PartitionedTableNoLock() because at this point the necessary
+	 * locks on the relation has already been acquired.
+	 */
+	if (PartitionedTableNoLock(relationId))
+	{
+		List *partitionList = PartitionList(relationId);
+		ShardInterval *shardInterval = LoadShardInterval(shardId);
+		int shardIndex = shardInterval->shardIndex;
+		ListCell *partitionCell = NULL;
+
+		foreach(partitionCell, partitionList)
+		{
+			Oid partitionOid = lfirst_oid(partitionCell);
+			uint64 partitionShardId = INVALID_SHARD_ID;
+			List *partitionShardPlacementList = NIL;
+			ShardPlacement *partitionInitialShardPlacement = NULL;
+
+			/*
+			 * During create_distributed_table, the partitions may not
+			 * have been created yet and so there are no placements yet.
+			 * We're already going to register them when we distribute
+			 * the partitions.
+			 */
+			if (!IsDistributedTable(partitionOid))
+			{
+				continue;
+			}
+
+			partitionShardId =
+				ColocatedShardIdInRelation(partitionOid, shardIndex);
+			partitionShardPlacementList = ShardPlacementList(partitionShardId);
+			partitionInitialShardPlacement =
+				(ShardPlacement *) linitial(partitionShardPlacementList);
+
+			/*
+			 * Recursively record all relation accesses of its partitions. Note that
+			 * we prefer to recursively call AssociatePlacementAccessWithRelation()
+			 * to support multi-level partitioned tables.
+			 */
+			AssociatePlacementAccessWithRelation(partitionInitialShardPlacement,
```

Comment:
What if we split this function into 2? One that takes a shardId and a wrapper that takes a placement. Then we can call the former recursively without having to look up placement metadata.